src/msg/async/EventEpoll.cc

```diff
@@ -117,12 +194,114 @@ int EpollDriver::resize_events(int newsize)
   return 0;
 }
 
+static inline unsigned int max_index_nr(struct epoll_uheader *header)
+{
+  return header->index_length >> 2;
+}
+
+static inline bool read_event(struct epoll_uheader *header, unsigned int *index,
+			      unsigned int idx, struct epoll_event *event)
+{
+  struct epoll_uitem *item;
+  unsigned int *item_idx_ptr;
+  unsigned int indeces_mask;
+
+  indeces_mask = max_index_nr(header) - 1;
+  if (indeces_mask & max_index_nr(header)) {
+    /* Should be pow2, corrupted header? */
+    return false;
+  }
+
+  item_idx_ptr = &index[idx & indeces_mask];
+
+  /*
+   * Spin here till we see valid index
+   */
+  while (!(idx = __atomic_load_n(item_idx_ptr, __ATOMIC_ACQUIRE)))
```

Comment:
if we use `atomic<unsigned> *` for `index`, then we could use
```c++
auto item_idx = &index[idx & indeces_mask];
while (!(idx = item_idx->load(std::memory_order_acquire)))
  ;
```

Reply:
I am not sure that this is a good idea to rely on a layout or internal implementation of an atomic<unsigned>.  Because if there is a version of C++ implementation, where atomic<unsigned> is bigger than 4b, or internally it does something differently - that will become a big problem. I would better stick with exactly raw atomic ops of C11 standard.